### ISSUE TITLE
`insertInitialDocument` option for document assign/merge

### ISSUE THREAD

**Issue by louisameline** [vuex-cloud/rfcs#16]
Feature request / reminder: it would be nice to have an option to insert the document on assign/merge too.

`dispatch('module/assign', { data: payload, parameters: { insertInitialDocument: true }}

**Comment by louisameline** [comment]
I have a question about the current behavior. It seems that I can't create a document when I'm in doc mode, because I can only "patch" it, which uses Firestore's `update`, which doesn't create the document if it's missing. It's annoying because I don't want to create a collection module just to create it, nor do I want to open a DBChannel for it.

So is there a reason not to use `set` with the `merge: true` option instead of `update`? It would create the document if it doesn't exist. Also, it's probably less work for the lib, as it doesn't require us to flatten the object and compute the individual property updates.

I know it's too late to change the current behavior as it would be a breaking change, but we might want to add the possibility to do it through an option. Or just keep it in mind for v2.

**Comment by mesqueeb** [comment]
I believe when I tested two years ago `set` with `merge: true` wasn't compatible with [batched writes](https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes).
And as you know, v-e-f 1.0 only uses batched writes, that's why I didn't use `set` with `merge: true`. For v2.0 I'm thinking to use batched writes as well so this will need some more testing.

The `insertInitialDoc` action is only relevant on "doc" mode in v-e-f 1.0, I haven't thought that much about "doc" mode in 2.0 yet, but if it is to be the same as a dynamically registered module via collection mode then it means it has to be _read_ first before it can be dynamically registered anyway, so we already know the doc's existence and contents.

The "insert" action on the other hand is executed on the "collection" module: `module/insert`, but the "assign" action is executed on that dynamically registered module itself: `module/abc123/assign`.

**Comment by louisameline** [comment]
> set with merge: true wasn't compatible with batched writes

Ok that would explain it. But the doc seems to say it's possible though https://firebase.google.com/docs/reference/js/firebase.firestore.WriteBatch#set

>  it has to be read first before it can be dynamically registered

Not necessarily, why? You may want to write to a file that you don't have read rights on, like logs. And you don't know if it exists yet.

My workaround with the current library (doc mode) when I don't know if the file already exists is to call `patch`, and if it fails with `error.code` === `not-found`, I call `insertInitialDoc` and then `update` again. But that's two extra calls that could have been avoided.

**Comment by mesqueeb** [comment]
>> it has to be read first before it can be dynamically registered

> Not necessarily, why?

Yeah, I meant, it will depend on how we implement it, but the way it was in my head originally is an explicit call to either `insert` or `get` or `stream` is what will dynamically register the documents. That's why I said "it has to be read first".

I think we'll just have to see if we can create a syntax that can dispatch a `merge` or `assign` to a document that hasn't been checked for existence yet.
With regular vuex you cannot dispatch an action to a module that's not been registered, so it'll depend on the implementation of the class I described in #11 I guess. 😉 

On the Firebase side we can definitely look into only using `set` with `merge: true`. If we can test and define the differences with their `update` action and we know what they are I don't see any reason not to.
